Feature: Search functionality for Shipyard/Outfitter Panels

Ember369 opened the pull request:
## Summary
Adds a search functionality to the shipyard/outfitter panels, similar to that of the map panel.

## Feature Details
The functionality of the search feature is consistent with the map panel equivalent (same dialog box and almost identical code except for the iteration through the ships/outfits). Pressing the `f` key opens up a dialog box allows the player to type a search term.
- If nothing matches, then nothing happens (the already selected ship/outfit is still selected, if one already is)
- If something is a match (the most similar partial or complete), the ship/outfit is selected and the panel is automatically scrolled so it is visible

Unlike the map panel equivalent, I haven't added a 'find' button in the UI, mostly because I couldn't find a good place to put it due to how busy it already is. I think it'd be fine to add UI in a separate PR later (if the community decides that it'd be a good addition), but personally I don't think there's enough room.

(btw I'm new to C++ and making PRs to ES, so not entirely sure if I did this right)

## Screenshots
![search shipyard](https://github.com/endless-sky/endless-sky/assets/59035469/983d49c0-4a98-4110-b9e3-bc6d040952a3)
![search outfitter](https://github.com/endless-sky/endless-sky/assets/59035469/ae51c052-00d0-4a75-9ccf-2defd4834bce)

### Automated Tests Added
N/A

## Performance Impact
N/A (I've tested it in Omnis and against monstrous plugins such as AGFFA to see if larger number of ships affect the search algorithm)

## Links
Assets: https://github.com/endless-sky/endless-sky-assets/pull/152
High-DPI: https://github.com/endless-sky/endless-sky-high-dpi/pull/386

Saugia commented:
It'd probably be a decent adjustment to the code, but any way to make this Ctrl-F instead of just F? This way it is also consistent with the Ctrl-F used on the map.

Ember369 commented:
Oh is it actually `Ctrl-F` on the map? As far as I can tell pressing `F` by itself seems to work fine?

bene-dictator commented:
I thunk it's just `F` on the map yeah

Ember369 commented:
Haha I guess this is just one of those weird ES things that aren't too well documented 😅

It wouldn't be particularly difficult to change both this PR and the existing map `F` search to `Ctrl+F` though (if we wanted to), since the code already exists in other features like grouping ships.

Saugia commented:
Oh, lol. As a habit from EV, I've never not done it without ctrl first, so forget that. F is fine.

TomGoodIdea commented:
Could you mention this feature in the shipyard / outfitter help messages? Especially if there's no button for it.

Ember369 commented:
Oh good idea, I'll work out how to do that 👍

Ember369 commented:
Alright I added 'find' help messages to the outfitter help message, and since there isn't an existing shipyard help message, I added one for that too. As a bonus I added it to the relevant map panels as well (the shipyard/outfitter panels, and the port panel), just for clarity.

TomGoodIdea commented on source/ShipyardPanel.cpp:
```diff
@@ -107,6 +107,12 @@ void ShipyardPanel::Step()
 {
 	ShopPanel::Step();
 	ShopPanel::CheckForMissions(Mission::SHIPYARD);
+	if(GetUI()->IsTop(this) && !checkedHelp)
+		// Use short-circuiting to only display one of them at a time.
+		// (The first valid condition encountered will make us skip the others.)
+		if(DoHelp("shipyard") || true)
+			// Either a help message was freshly displayed, or all of them have already been seen.
+			checkedHelp = true;
```
You don't need this if there's only one help message. Idk if the IsTop condition should stay.
```suggestion
	if(GetUI()->IsTop(this))
		DoHelp("shipyard");
```

TomGoodIdea commented on source/ShipyardPanel.h:
```diff
@@ -63,6 +63,9 @@ class ShipyardPanel : public ShopPanel {
 	int modifier;
 
 	Sale<Ship> shipyard;
+
+	// Keep track of how many of the outfitter help screens have been shown
+	bool checkedHelp = false;
```
```suggestion
```

Ember369 commented:
Good catch @TomGoodIdea, I've committed your suggestions

TomGoodIdea commented on source/ShopPanel.cpp:
```diff
@@ -616,6 +619,50 @@ bool ShopPanel::Scroll(double dx, double dy)
+	{
+		const Ship *ship = it->GetShip();
+		const Outfit *outfit = it->GetOutfit();
+		if(ship && HasItem(ship->VariantName()))
+			index = Search(ship->VariantName(), text);
+		else if(outfit && HasItem(outfit->TrueName()))
+			index = Search(outfit->TrueName(), text);
```
Doesn't `HasItem` check only if something is for sale?

Also, variant / true names are internal, the player sees only display names.

Ember369 replied:
Ahh good catch for the display model names. I've changed the `Search()` functions to use `ship->DisplayModelName()` and `outfit->DisplayName()` respectively.

In regards to what `HasItem()` checks, for ships it only searches for ships on sale (since they're the only ones that appear in the shop view), and for outfits it searches for all that are on sale, in cargo, in storage, or installed. Should that be sufficient?

Koranir commented:
I think you should add a button that activates this feature. Slightly to the left of the scroll down button should do nicely.

If you need, I can make a PR on your repo to add in the button.

Ember369 commented:
Hmm around here you reckon? You're right, it's not as obstructing as I thought it would be.

![find button](https://github.com/endless-sky/endless-sky/assets/59035469/e41ee14d-30e3-428f-9a9f-5117cffc7157)

Ember369 commented on source/ShopPanel.cpp:
```diff
@@ -1413,6 +1467,10 @@ vector<ShopPanel::Zone>::const_iterator ShopPanel::Selected() const
 // letter of the button (or ' ' if it's not on a button).
 char ShopPanel::CheckButton(int x, int y)
 {
+	if(x > Screen::Right() - SIDEBAR_WIDTH - 380 && x < Screen::Right() - SIDEBAR_WIDTH - 320 &&
```
The `ShopPanel::CheckButton()` function begins by ignoring all points outside the "button zone", so I had to add the find button clicking functionality to before this part. Since it's separate from the other button code, it's kind of ugly imo.

Koranir replied:
Looks fine to me.

Koranir commented:
Hmm... I was thinking more of a 🔎 kind of icon, like the sorting buttons on the missions screen.

Ember369 commented:
Ahh, so something more like this? (I copy and pasted the spyglass emoji from above, so we'll eventually need our own)

![find button 2](https://github.com/endless-sky/endless-sky/assets/59035469/cc7c7d38-33d5-43cd-9e94-abdd9e8bfe60)

Koranir commented:
Yup.

Ember369 commented:
Well this is what we ended up deciding on, in the Discord server. I have the relevant `.svg` files for the images, so whenever (or if) this PR gets the green flag I can make a PR on the ES assets repo
![rounded rectangle version](https://github.com/endless-sky/endless-sky/assets/59035469/e406db85-a6f6-43f7-815b-df34ed3d07cd)

Amazinite commented:
> so whenever (or if) this PR gets the green flag I can make a PR on the ES assets repo.

It's usually better to open high DPI and asset PRs beforehand so that if the PR looks good, we can merge it without needing to wait on those other PRs.

Amazinite commented on images/ui/find icon light.png:
Having the entire button become lighter when you're hovering over it doesn't match the style of the rest of the game. For other buttons, the text becomes lighter, but the background of the button remains the same. The analogous interaction here would be to have only the magnifying glass icon become lighter instead of causing the entire button to become lighter.

Amazinite commented on source/ShopPanel.cpp:
```diff
@@ -1413,6 +1466,10 @@ vector<ShopPanel::Zone>::const_iterator ShopPanel::Selected() const
 // letter of the button (or ' ' if it's not on a button).
 char ShopPanel::CheckButton(int x, int y)
 {
+	if(x > Screen::Right() - SIDEBAR_WIDTH - 342 && x < Screen::Right() - SIDEBAR_WIDTH - 316 &&
+		y > Screen::Bottom() - 31 && y < Screen::Bottom() - 4)
+		return 'f';
```
We need to make this use interface data at some point. This is ugly. 🙈
You're fine to leave it as is now, just a problem for future us.

Ember369 replied:
Yeah I mentioned the uglyness of this above. The current code just isn't designed to have buttons on the left side of the screen 😬

Amazinite commented on source/ShopPanel.cpp:
```diff
@@ -433,6 +433,9 @@ bool ShopPanel::KeyDown(SDL_Keycode key, Uint16 mod, const Command &command, boo
 	else if(key == SDLK_TAB)
 		activePane = (activePane == ShopPane::Main ? ShopPane::Sidebar : ShopPane::Main);
+	else if(key == 'f')
+		GetUI()->Push(new Dialog(
+			this, &ShopPanel::DoFind, "Search for:"));
```
I see that the map find adds a new line here, but that's also a super old line of code; it's not really a part of our current conventions to add a new line unless it's for aesthetic purposes (which this doesn't seem to be to me) or because the line is too long (which this isn't).
```suggestion
		GetUI()->Push(new Dialog(this, &ShopPanel::DoFind, "Search for:"));
```

Amazinite commented on source/ShopPanel.cpp:
```diff
@@ -616,6 +619,50 @@ bool ShopPanel::Scroll(double dx, double dy)
+	int index;
+	for( ; it != zones.end(); ++it)
+	{
+		const Ship *ship = it->GetShip();
+		const Outfit *outfit = it->GetOutfit();
+		if(ship && HasItem(ship->TrueModelName()))
+			index = Search(ship->DisplayModelName(), text);
+		else if(outfit && HasItem(outfit->TrueName()))
+			index = Search(outfit->DisplayName(), text);
```
We should probably create a FindItem function that ShipyardPanel and OutfitterPanel implement that gets called here, just as MapSalesPanel/MapShipyardPanel/MapOutfitterPanel do.

https://github.com/endless-sky/endless-sky/blob/1664c421cf2083c49cb9d7361cb06ae671c93d82/source/MapSalesPanel.cpp#L417-L426

https://github.com/endless-sky/endless-sky/blob/1664c421cf2083c49cb9d7361cb06ae671c93d82/source/MapShipyardPanel.cpp#L173-L189

https://github.com/endless-sky/endless-sky/blob/1664c421cf2083c49cb9d7361cb06ae671c93d82/source/MapOutfitterPanel.cpp#L164-L180

Ember369 replied:
What'd be the best way to do this? I wanted to make a FindItem function separate to `ShopPanel`, but `ShipyardPanel` and `OutfitterPanel` don't have access to any of the zones that require iterating through (which is why I left it where it currently is). I'm not sure what the best approach is from a code style perspective

Amazinite replied:
> ShipyardPanel and OutfitterPanel don't have access to any of the zones

They don't? The `zones` variable is protected, not private.

Koranir replied:
Fyi, `protected` variables are accessible from child classes.

Ember369 replied:
Yup you were both right; I read the code wrong. I've split the functionality into the child classes now 👍

Ember369 commented:
Well I reckon the UI's more consistent now
![more consistent](https://github.com/endless-sky/endless-sky/assets/59035469/87fa20c1-65de-48a8-ba13-3bb70d5430e4)

Amazinite commented on source/OutfitterPanel.cpp:
```diff
@@ -980,3 +980,26 @@ const vector<Ship *> OutfitterPanel::GetShipsToOutfit(bool isBuy) const
+	int bestIndex = 9999;
+	int bestItem = -1;
+	vector<Zone>::const_iterator it = zones.begin();
+	for(unsigned int i = 0; i < zones.size(); ++i)
```
You can chain multiple iterators together in a for statement.
```suggestion
	for(unsigned int i = 0; i < zones.size(); ++i, ++it)
```

Ember369 replied:
Oh awesome, I didn't realise this was possible

Amazinite commented on source/OutfitterPanel.cpp:
```diff
@@ -980,3 +980,26 @@ const vector<Ship *> OutfitterPanel::GetShipsToOutfit(bool isBuy) const
+			if(!index)
+				return i;
+		}
+		++it;
```
```suggestion
```

Amazinite commented on source/ShipyardPanel.cpp:
```diff
@@ -401,3 +403,24 @@ void ShipyardPanel::SellShip()
+	int bestIndex = 9999;
+	int bestItem = -1;
+	vector<Zone>::const_iterator it = zones.begin();
+	for(unsigned int i = 0; i < zones.size(); ++i)
```
```suggestion
	for(unsigned int i = 0; i < zones.size(); ++i, ++it)
```

Amazinite commented on source/ShipyardPanel.cpp:
```diff
@@ -401,3 +403,24 @@ void ShipyardPanel::SellShip()
+			if(!index)
+				return i;
+		}
+		++it;
```
```suggestion
```

Amazinite commented on source/OutfitterPanel.cpp:
```diff
@@ -980,3 +980,26 @@ const vector<Ship *> OutfitterPanel::GetShipsToOutfit(bool isBuy) const
+{
+	int bestIndex = 9999;
+	int bestItem = -1;
+	vector<Zone>::const_iterator it = zones.begin();
```
We usually just use `auto` for iterators.
```suggestion
	auto it = zones.begin();
```

Amazinite commented on source/ShipyardPanel.cpp:
```diff
@@ -401,3 +403,24 @@ void ShipyardPanel::SellShip()
+{
+	int bestIndex = 9999;
+	int bestItem = -1;
+	vector<Zone>::const_iterator it = zones.begin();
```
```suggestion
	auto it = zones.begin();
```

Amazinite commented on source/ShopPanel.cpp:
```diff
@@ -616,6 +618,32 @@ bool ShopPanel::Scroll(double dx, double dy)
+	int index = FindItem(text);
+	if(index >= 0 && index < static_cast<int>(zones.size()))
+	{
+		vector<Zone>::const_iterator best = std::next(zones.begin(), index);
```
```suggestion
		auto best = std::next(zones.begin(), index);
```

Amazinite commented on source/ShopPanel.cpp:
```diff
@@ -616,6 +618,32 @@ bool ShopPanel::Scroll(double dx, double dy)
+			selectedOutfit = best->GetOutfit();
+		MainAutoScroll(best);
+	}
+	return;
```
No need to return at the end of a void function.
```suggestion
```

Amazinite commented on source/ShopPanel.cpp:
```diff
@@ -616,6 +618,32 @@ bool ShopPanel::Scroll(double dx, double dy)
+
+
+
+int ShopPanel::Search(const string &str, const string &sub)
+{
+	auto it = search(str.begin(), str.end(), sub.begin(), sub.end(),
+		[](char a, char b) { return toupper(a) == toupper(b); });
+	return (it == str.end() ? -1 : it - str.begin());
+}
```
Since this is just the same Search function as is used in MapPanel, we should probably make it a static function of some other class. Perhaps put it in Format, which I believe all these panels already have access to.

Ember369 replied:
Ahh alright, done

Amazinite commented:
Still need the 2x and assets PRs before this can be merged.

Ember369 commented:
Just made the asset and high-dpi PRs 👍

Amazinite commented:
:+1:
